compiler/src/toEC.ml

```diff
@@ -357,156 +357,206 @@ type ec_prog = ec_item list
-          aux
-      end
-  in
-  List.map do1 tys
+      let rec aux i =
+        let s = Format.sprintf "%s_%i" s i in
+        if Ss.mem s !(env.alls) then aux (i+1)
+        else s in
+      aux 0
+
+  let mkname env n =
+    n |> String.uncapitalize_ascii |> escape |> create_name env
+
+  let set_var env x =
+    let s = mkname env x.v_name in
+    { env with
+      alls = ref (Ss.add s !(env.alls));
+      vars = Mv.add x s env.vars }
+
+  let add_ty env = function
+      | Bty _ -> ()
+      | Arr (_ws, n) -> add_Array env n
+
+  let empty arch pd array_theories =
+    {
+      arch;
+      pd;
+      alls = ref keywords;
+      vars = Mv.empty;
+      glob = Ms.empty;
+      funs = Mf.empty;
+      array_theories;
+      auxv  = ref Mpty.empty;
+      count = Mpty.empty;
+      randombytes = ref Sint.empty;
+    }
+
+  let set_fun env fd =
+    let s = mkname env fd.f_name.fn_name in
+    let funs =
+      Mf.add fd.f_name (s, ((*mk_tys*) fd.f_tyout, (*mk_tys*)fd.f_tyin)) env.funs in
+    { env with funs; alls = ref (Ss.add s !(env.alls)) }
+
+  let get_funtype env f = snd (Mf.find f env.funs)
+
+  let get_funname env f = fst (Mf.find f env.funs)
+
+  (*
+    Auxiliary variables created by "create_aux" have the given prefix and their
+    name, and are declared with the given type. Each created variable is
+    guaranteed to be unique for all create_aux calls with **the same env**
+    and (recursively) with **envs further derived by new_aux_range**.
+    However, aux var may be resued across other env (e.g. in two sibling
+    envs created by two calls to new_aux_range on the same env).
+
+    This is implemented by keeping a per-env count of created (prefix, ty) auxs
+    (env.count), while env.auxv tracks the complete list of created aux in the
+    whole function (for re-used and initial declaration).
+    new_aux_range copies env.count, ensuring that we don't reuse variables
+    already created for this env, but that different calls to new_aux_range do
+    not share the same env.count (hence may use the same auxs).
+  *)
+  let create_aux env prefix ty =
+    let i = try Mpty.find (prefix, ty) env.count with Not_found -> 0 in
+    let l = try Mpty.find (prefix, ty) !(env.auxv) with Not_found -> BatVect.empty in
+    env.count <- Mpty.add (prefix,ty) (i+1) env.count;
+    if i < BatVect.length l then begin
+      BatVect.get l i
+    end else begin
+      let aux = create_name env prefix in
+      env.auxv := Mpty.add (prefix, ty) (BatVect.append aux l) !(env.auxv);
+      env.alls := Ss.add aux !(env.alls);
+      aux
+    end
 
-let pop_aux env prefix ty =
-  let i = try Mty.find (prefix, ty) !(env.count) with Not_found -> 0 in
-  if 0 < i then
-    env.count := Mty.add (prefix,ty) (i-1) !(env.count)
+  (* Return the last created aux for (prefix, ty) in this env. *)
+  let reuse_aux env prefix ty =
+    let i = Mpty.find (prefix, ty) env.count in
+    let l = Mpty.find (prefix, ty) !(env.auxv) in
+    BatVect.get l (i-1)
```

Comment:
It might be interesting to add a test for the size of the list `l`, and if the list is empty a call to `create_aux`.

Reply:
It('s intentionally not done: if `reuse_aux` is called is called and `create_aux` has not been called before, then it's a bug. It's then better to crash than to (probably) generate incorrect output.

Reply:
I am agree with Gaetan

Reply:
Lyonel are you agree to remove this change request ?